pytensor/graph/rewriting/basic.py

```diff
@@ -743,14 +743,20 @@ def apply(self, fgraph):
                         ):
                             continue
 
-                if len(pairs) == 1 and pairs[0][0].type != pairs[0][1].type:
-                    res = pairs[0][0].type.convert_variable(pairs[0][1])
-
-                    # Since the fgraph.replace only checks the convert_variable
-                    # in one way, we change the order in the case that
-                    # convert_variable will not be successful.
-                    if not res:
-                        pairs = [(pairs[0][1], pairs[0][0])]
+                # Replace by more specific types
```

Comment:
What is the more generic context here? What the function is doing? (Add comments)

Reply:
I'll add a comment, but it may help to just read what the rewriter does: https://github.com/pymc-devs/pytensor/blob/71e3ec7adb7b214a4d25c5c1b2a6a11f0108d7ec/pytensor/graph/rewriting/basic.py#L667-L679

The thing is that two equivalent nodes may have been initially created with different output types (ie. you had more precise static type info in one of the branches before they were rendered  equivalently). It's a bit tricky to come up with an example but it definitely occurred in https://github.com/pymc-labs/pymc-marketing/issues/332